File: schemas/pack-5/validate-fixtures.py
from pathlib import Path
from datetime import datetime
import json
import re

root = Path(__file__).resolve().parent
schema_path = root / "event.schema.json"
schema = json.loads(schema_path.read_text(encoding="utf-8"))

if schema.get("$schema") != "https://json-schema.org/draft/2020-12/schema":
    raise ValueError("Schema does not declare JSON Schema Draft 2020-12")

def resolve_ref(reference):
    if not reference.startswith("#/"):
        raise ValueError(f"Unsupported external reference: {reference}")
    value = schema
    for part in reference[2:].split("/"):
        value = value[part.replace("~1", "/").replace("~0", "~")]
    return value

def validate(rule, value, path="$"):
    errors = []
    if "$ref" in rule:
        return validate(resolve_ref(rule["$ref"]), value, path)
    if "required" in rule and isinstance(value, dict):
        for name in rule["required"]:
            if name not in value:
                errors.append(f"{path}.{name}: required property missing")
    if "const" in rule and value != rule["const"]:
        errors.append(f"{path}: must equal {rule['const']!r}")
    if "enum" in rule and value not in rule["enum"]:
        errors.append(f"{path}: value is not in the permitted catalogue")
    expected_type = rule.get("type")
    if expected_type == "object":
        if not isinstance(value, dict):
            return errors + [f"{path}: must be an object"]
        if rule.get("additionalProperties") is False:
            allowed = set(rule.get("properties", {}))
            for name in value:
                if name not in allowed:
                    errors.append(f"{path}.{name}: additional property not allowed")
        for name, child in value.items():
            if name in rule.get("properties", {}):
                errors.extend(validate(rule["properties"][name], child, f"{path}.{name}"))
    elif expected_type == "array":
        if not isinstance(value, list):
            return errors + [f"{path}: must be an array"]
        if len(value) < rule.get("minItems", 0):
            errors.append(f"{path}: contains too few items")
        if rule.get("uniqueItems") and len({json.dumps(item, sort_keys=True) for item in value}) != len(value):
            errors.append(f"{path}: items must be unique")
        for index, child in enumerate(value):
            errors.extend(validate(rule.get("items", {}), child, f"{path}[{index}]"))
    elif expected_type == "string":
        if not isinstance(value, str):
            return errors + [f"{path}: must be a string"]
        if len(value) < rule.get("minLength", 0):
            errors.append(f"{path}: string is too short")
        if len(value) > rule.get("maxLength", len(value)):
            errors.append(f"{path}: string is too long")
        if "pattern" in rule and not re.search(rule["pattern"], value):
            errors.append(f"{path}: string does not match required pattern")
        if rule.get("format") == "date-time":
            try:
                datetime.fromisoformat(value.replace("Z", "+00:00"))
            except ValueError:
                errors.append(f"{path}: invalid ISO 8601 date-time")
    elif expected_type == "integer":
        if not isinstance(value, int) or isinstance(value, bool):
            return errors + [f"{path}: must be an integer"]
        if value < rule.get("minimum", value):
            errors.append(f"{path}: below minimum")
    for branch_set in ("anyOf", "oneOf"):
        if branch_set in rule:
            passes = sum(not validate(branch, value, path) for branch in rule[branch_set])
            required_passes = 1
            if (branch_set == "anyOf" and passes < required_passes) or (branch_set == "oneOf" and passes != required_passes):
                errors.append(f"{path}: does not satisfy {branch_set}")
    return errors

results = []
failures = 0
for expected, folder in (("pass", "valid"), ("fail", "invalid")):
    for fixture_path in sorted((root / "fixtures" / folder).glob("*.json")):
        instance = json.loads(fixture_path.read_text(encoding="utf-8"))
        errors = validate(schema, instance)
        actual = "fail" if errors else "pass"
        passed = actual == expected
        failures += 0 if passed else 1
        results.append({
            "schema": schema_path.name,
            "fixture": fixture_path.name,
            "expected": expected,
            "actual": actual,
            "passed": passed,
            "errors": errors,
        })

report = {
    "validator": "local deterministic validator for the Draft 2020-12 keywords used by this pack",
    "schemaSyntaxValid": True,
    "failures": failures,
    "results": results,
}
(root / "schema-validation-report.json").write_text(
    json.dumps(report, indent=2) + "\n", encoding="utf-8"
)
print(json.dumps(report, indent=2))
raise SystemExit(1 if failures else 0)

File: .codex-staging/pack-2-schema-draft/02 Pack 2/Pack 2 Schema Draft/scripts/validate-fixtures.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const root = path.resolve(__dirname, '..');
function readJson(file) { return JSON.parse(fs.readFileSync(file, 'utf8')); }
function isObject(v) { return v && typeof v === 'object' && !Array.isArray(v); }
function typeOk(expected, value) {
  const list = Array.isArray(expected) ? expected : [expected];
  return list.some(t => t === 'array' ? Array.isArray(value) : t === 'object' ? isObject(value) : t === 'integer' ? Number.isInteger(value) : t === 'null' ? value === null : typeof value === t);
}
function resolveRef(schema, ref) {
  if (!ref.startsWith('#/$defs/')) throw new Error(`Unsupported ref ${ref}`);
  return schema.$defs[ref.slice('#/$defs/'.length)];
}
function validateNode(rootSchema, schema, value, pointer, errors) {
  if (schema.$ref) return validateNode(rootSchema, resolveRef(rootSchema, schema.$ref), value, pointer, errors);
  if (schema.const !== undefined && value !== schema.const) errors.push(`${pointer}: expected const ${schema.const}`);
  if (schema.type && !typeOk(schema.type, value)) { errors.push(`${pointer}: type mismatch`); return; }
  if (schema.enum && !schema.enum.includes(value)) errors.push(`${pointer}: value not in enum`);
  if (schema.minLength !== undefined && typeof value === 'string' && value.length < schema.minLength) errors.push(`${pointer}: below minLength`);
  if (schema.minimum !== undefined && typeof value === 'number' && value < schema.minimum) errors.push(`${pointer}: below minimum`);
  if (schema.pattern && typeof value === 'string' && !(new RegExp(schema.pattern).test(value))) errors.push(`${pointer}: pattern mismatch`);
  if (schema.format === 'date-time' && typeof value === 'string' && Number.isNaN(Date.parse(value))) errors.push(`${pointer}: invalid date-time`);
  if (schema.type === 'object' && isObject(value)) {
    const props = schema.properties || {};
    for (const req of schema.required || []) if (!(req in value)) errors.push(`${pointer}/${req}: required property missing`);
    if (schema.additionalProperties === false) for (const key of Object.keys(value)) if (!(key in props)) errors.push(`${pointer}/${key}: additional property not allowed`);
    for (const [key, child] of Object.entries(props)) if (key in value) validateNode(rootSchema, child, value[key], `${pointer}/${key}`, errors);
  }
  if (schema.type === 'array' && Array.isArray(value)) {
    if (schema.minItems !== undefined && value.length < schema.minItems) errors.push(`${pointer}: below minItems`);
    if (schema.items) value.forEach((item, i) => validateNode(rootSchema, schema.items, item, `${pointer}/${i}`, errors));
  }
}
function validate(schema, data) { const errors = []; validateNode(schema, schema, data, '', errors); return errors; }

let failures = 0;
const results = [];
const schemas = fs.readdirSync(path.join(root, 'schemas')).filter(f => f.endsWith('.schema.json')).sort();
for (const schemaFile of schemas) {
  const base = schemaFile.replace('.schema.json', '');
  const schema = readJson(path.join(root, 'schemas', schemaFile));
  for (const [kind, shouldPass] of [['valid', true], ['invalid', false]]) {
    const fixture = `${kind}-${base}.json`;
    const data = readJson(path.join(root, 'fixtures', kind, fixture));
    const errors = validate(schema, data);
    const actualPass = errors.length === 0;
    if (actualPass !== shouldPass) failures++;
    results.push({schema: schemaFile, fixture, expected: shouldPass ? 'pass' : 'fail', actual: actualPass ? 'pass' : 'fail', errors});
  }
}
const output = {validator: 'local structural JSON Schema subset validator for Pack 2 fixtures', failures, results};
console.log(JSON.stringify(output, null, 2));
process.exit(failures ? 1 : 0);
